[BrutBridge] New bridge

VerifiedJoseph opened the pull request:
A bridge for [Brut](https://www.brut.media/us). Returns the 5 newest videos by category and edition. closes #1092

VerifiedJoseph committed:
Create BrutBridge.php

VerifiedJoseph committed:
Set cache timeout to 1800 seconds (30 mins)

logmanoriginal commented on bridges/BrutBridge.php:
```diff
@@ -0,0 +1,128 @@
+
+			$videoImage = $videoPageHtml->find("meta[name=twitter:image]")[0]->content;
+			
+			preg_match($this->videoImageRegex, $videoImage, $matches)
```
This operation fails if the file extension is `png` instead of `img`.
I strongly suggest parsing the image URL manually.
This works for me.

```php
/**
 * Extract video ID + type from twitter image
 *
 * Example (wrapped):
 *  https://img.brut.media/thumbnail/
 *  the-life-of-rita-moreno-2cce75b5-d448-44d2-a97c-ca50d6470dd4-square.jpg
 *  ?ts=1559337892
 */
$twimg = $videoPageHtml->find('meta[name="twitter:image"]', 0);
$fpath = parse_url($twimg->content, PHP_URL_PATH);
$fname = basename($fpath);
$fname = substr($fname, 0, strrpos($fname, '.'));
$parts = explode('-', $fname);

$videoId = implode('-', array_splice($parts, -6, 5));
$videoType = end($parts);
```

It doesn't rely on the URL, file extension or query parameters and therefore should work as long as the ID and video type are the last six parts of the file name.

VerifiedJoseph replied:
This is much better than my regex mess. I've had to make a small change so it works with URLs like `https://img.brut.media/thumbnail/restaurant-fights-food-waste-by-lowering-prices-f80d9f94-2c08-4187-a9ab-2788befeebbb-square-auto.jpg`. 

```PHP
$twimg = $videoPageHtml->find('meta[name="twitter:image"]', 0);
$fpath = parse_url($twimg->content, PHP_URL_PATH);
$fname = basename($fpath);
$fname = substr($fname, 0, strrpos($fname, '.'));
$parts = explode('-', $fname);

if (end($parts) === 'auto') {
	$key = array_search('auto', $parts);
	unset($parts[$key]);
}

$videoId = implode('-', array_splice($parts, -6, 5));
$videoType = end($parts);
```

Thanks for the feedback.

logmanoriginal commented on bridges/BrutBridge.php:
```diff
@@ -0,0 +1,128 @@
+
+	public function collectData() {
+
+		$this->feedURI = self::URI . '/' . $this->getInput('edition') . '/' . $this->getInput('category');
```
Instead of storing the URL in a private variable, simply build the URL in `getURI` and call the function here. Have a look at how other bridges implement this.

For example Twitter:

https://github.com/RSS-Bridge/rss-bridge/blob/3d231a417f3670d3e93fb63e9af9d62a5d17e87d/bridges/TwitterBridge.php#L126-L145

https://github.com/RSS-Bridge/rss-bridge/blob/3d231a417f3670d3e93fb63e9af9d62a5d17e87d/bridges/TwitterBridge.php#L150

This will also ensure that the correct URL is used for the HTML preview (uses `getURI` for that).

logmanoriginal commented on bridges/BrutBridge.php:
```diff
@@ -0,0 +1,128 @@
+
+	public function getURI() {
+
+		if ($this->feedURI) {
```
See above.

logmanoriginal commented on bridges/BrutBridge.php:
```diff
@@ -0,0 +1,128 @@
+		return strtotime($description->find('div.date', 0)->innertext);
+	}
+
+	private function processConent($description, $videoId, $videoType, $videoImage) {
```
Typo `processConent` => `processContent`

logmanoriginal reviewed (changes requested):
Interesting bridge, thanks for the PR :+1: 

Find below a few comments for consideration. Please also take a look at the complaints from Travis.

VerifiedJoseph committed:
Fix function name typo (processConent => processContent)

VerifiedJoseph committed:
Use getURI() to create URLs

Use getURI() to create URLs as suggested by @LogMANOriginal in code review.

https://github.com/RSS-Bridge/rss-bridge/pull/1159#discussion_r291721457

VerifiedJoseph committed:
Replace twitter image regex, add processTwitterImage()

Replaces twitter image regex with code suggested by @LogMANOriginal in code review.

https://github.com/RSS-Bridge/rss-bridge/pull/1159#discussion_r291720090

VerifiedJoseph committed:
Fix travis-ci issues

VerifiedJoseph committed:
Fix more travis-ci issues

logmanoriginal commented on bridges/BrutBridge.php:
```diff
@@ -0,0 +1,142 @@
+			$item['timestamp'] = $this->processDate($description);
+			$item['enclosures'][] = $this->videoImage;
+
+			$this->items[] = $item;
```
One more detail. To limit the number of items add a check right after adding the items instead of checking at the top of the loop. This will break the loop as soon as possible, skipping the next iteration entirely.

```suggestion
			$this->items[] = $item;
			if (count($this->items) >= 5)
				break;
```

It also doesn't rely on the array index, which could change in the future.

VerifiedJoseph committed:
Move location of item limit check

Move location of item limit check and use code suggested by @LogMANOriginal in code review.

https://github.com/RSS-Bridge/rss-bridge/pull/1159#discussion_r291807974

VerifiedJoseph committed:
Fix travis-ci issue

Remove whitespace from end of line 81

logmanoriginal commented:
Looks good to me. Thanks again for the PR :+1:

dvikan commented:
@VerifiedJoseph The bridge is currently broken.
